### PR TITLE
[HigherOrderOp] track input mutations via tensor version

### LINKED ISSUE #115277: HigherOrderOp graph capturing is wrong for buffer mutation
### 🐛 Describe the bug

```python
class M(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.register_buffer("buffer", torch.randn(4, 4))

    def forward(self, x):
        def true_fn(x):
            self.buffer.add_(5)
            return x.cos() + self.buffer.sum()

        def false_fn(x):
            return x.sin()

        a = torch.cond(x.shape[0] > 4, true_fn, false_fn, [x])
        return (a + 3, a + 4)


inp = torch.randn(3, 4)
ep = torch.export.export(M(), (inp,))
print(ep.graph_module.graph)
print(ep.graph_module.true_graph_0.graph)
```
prints:
```
graph():
    %arg0_1 : [num_users=1] = placeholder[target=arg0_1]
    %l_x_ : [num_users=1] = placeholder[target=l_x_]
    %true_graph_0 : [num_users=1] = get_attr[target=true_graph_0]
    %false_graph_0 : [num_users=1] = get_attr[target=false_graph_0]
    %conditional : [num_users=1] = call_function[target=torch.ops.higher_order.cond](args = (False, %true_graph_0, %false_graph_0, [%l_x_, %arg0_1]), kwargs = {})
    %getitem : [num_users=2] = call_function[target=operator.getitem](args = (%conditional, 0), kwargs = {})
    %add : [num_users=1] = call_function[target=torch.ops.aten.add.Tensor](args = (%getitem, 3), kwargs = {})
    %add_1 : [num_users=1] = call_function[target=torch.ops.aten.add.Tensor](args = (%getitem, 4), kwargs = {})
    return (add, add_1)
graph():
    %arg0_1 : [num_users=1] = placeholder[target=arg0_1]
    %arg1_1 : [num_users=1] = placeholder[target=arg1_1]
    %add : [num_users=1] = call_function[target=torch.ops.aten.add.Tensor](args = (%arg1_1, 5), kwargs = {})
    %cos : [num_users=1] = call_function[target=torch.ops.aten.cos.default](args = (%arg0_1,), kwargs = {})
    %sum_1 : [num_users=1] = call_function[target=torch.ops.aten.sum.default](args = (%add,), kwargs = {})
    %add_1 : [num_users=1] = call_function[target=torch.ops.aten.add.Tensor](args = (%cos, %sum_1), kwargs = {})
    return (add_1,)
 ```
 As you can see true_graph is not returning %add as extra output. It should because this refers to a mutated buffer which needs to be propagated to the top level graph. 

### Versions

main

### PR DESCRIPTION
**Pull request by ydwu4**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #115451

Fixes #115277 by banning tensor buffer, tensor input and tensor lifted clousures mutations and requiring users to rewrite their code for now. It implements the idea of using tensor._version to detect mutations from @zou3519. After this PR, speculate_subgraph will return input proxies that gets mutated. Different HOOs can choose different strategies to make use of this info.

We might be able to support these mutations for HOOs but need to wait for user requests and decide.

Test Plan:
Added new tests for cond. Also modify existing tests that aliasing inputs.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

Differential Revision: [D52096595](https://our.internmc.facebook.com/intern/diff/D52096595)

### CODE REVIEW

File: torch/_dynamo/output_graph.py [hunk 1]
```diff
@@ -1829,6 +1842,19 @@ def maybe_lift_tracked_freevar_to_input(self, arg):
             return arg
         return self.lift_tracked_freevar_to_input(arg)
 
+    def track_mutations_for_tensor_inputs(self, proxies: List[fx.Proxy]):
+        self.tensor_input_versions.update(
```
Comment (zou3519) [inline torch/_dynamo/output_graph.py:1846]: The version counter logic gets disabled when torch.inference_mode is active. I don't think this gets used often, but I could be wrong. Could we:
- not track this information when torch.inference_mode is active
- raise unimplemented in cond if torch.inference_mode is active

Comment (zou3519) [reply torch/_dynamo/output_graph.py:1846]: Actually, I'm not sure that torch.inference_mode isn't used often.

Comment (zou3519) [reply torch/_dynamo/output_graph.py:1846]: Proposal #2, that doesn't use version counters:
- once we have captured a graph, we run clones of the input FakeTensors through the graph with a torch_dispatch mode active
- the torch_dispatch mode will raise an error if any inputs are being mutated. We can tell this by looking at the schema of the operators.

Downsides: doesn't really work with subclasses

**Comment by ydwu4** [comment]
@ydwu4 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D52096595).

File: torch/_dynamo/variables/higher_order_ops.py [hunk 2]
```diff
@@ -116,6 +116,41 @@ def _call_function_and_unflatten_output(tx, fn, args, kwargs, ret_vt, ret_treesp
     )
 
 
+# NOTE:
+# Mutation detection: We detect mutations by checking the example_value's _version
+# This approach has a corner case that if torch.compile is wrapped
+# by an inference_mode(True) context manager, the tensor version will not be upated.
```
Comment (zou3519) [inline torch/_dynamo/variables/higher_order_ops.py:122]: Can we add some tests that cond still errors out in a situation where it should if it is under torch.inference_mode?

File: torch/_dynamo/variables/higher_order_ops.py [hunk 3]
```diff
@@ -661,30 +696,63 @@ def _insert_or_replace_phs(new_args, name_suffix):
...
+        for branch_name, module in (
+            (True, true_graph_module),
+            (False, false_graph_module),
+        ):
+            try:
+                mutated_inputs, aliased_inputs = _detect_input_mutations_and_aliasing(
+                    tx, module, combined_args
+                )
+            except Exception as e:
+                unimplemented(
```
Comment (zou3519) [inline torch/_dynamo/variables/higher_order_ops.py:735]: try-except with bare Exception is dangerous, do we need this?

**Review by zou3519 — COMMENTED** [review]
Code looks good to me, but we should run a test case under inference_mode for completeness

File: torch/_dynamo/variables/higher_order_ops.py (hunk 3 above)
Comment (ydwu4) [reply torch/_dynamo/variables/higher_order_ops.py:735]: Yeah, we can be more specific about the exception.

**Comment by ydwu4** [comment]
Tests added!

**Pull request closed without merging**